File: apps/mobile/lib/state.dart
// state.dart — estado global con ChangeNotifier + provider:
//   PrefsState  → tema y servidor (SharedPreferences; nunca credenciales)
//   AuthState   → sesión (API key "Lixbon Mobile" en el Keystore vía
//                 flutter_secure_storage) — se rota en cada login
//   ChatState   → modelos, conversación activa y streaming
import 'dart:async';
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:shared_preferences/shared_preferences.dart';

import 'api.dart';
import 'oauth.dart';
import 'sse.dart';

const kDefaultApiBase = 'https://lixbon.com';
const _kSecureKey = 'lixbon_api_key';
const _kMobileKeyName = 'Lixbon Mobile';

// ── Preferencias locales ───────────────────────────────────────────────────

class PrefsState extends ChangeNotifier {
  ThemeMode themeMode = ThemeMode.system;
  String apiBase = kDefaultApiBase;

  Future<void> hydrate() async {
    final prefs = await SharedPreferences.getInstance();
    final mode = prefs.getString('themeMode');
    themeMode = switch (mode) {
      'light' => ThemeMode.light,
      'dark' => ThemeMode.dark,
      _ => ThemeMode.system,
    };
    apiBase = prefs.getString('apiBase') ?? kDefaultApiBase;
    notifyListeners();
  }

  Future<void> setThemeMode(ThemeMode mode) async {
    themeMode = mode;
    notifyListeners();
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('themeMode', switch (mode) {
      ThemeMode.light => 'light',
      ThemeMode.dark => 'dark',
      ThemeMode.system => 'system',
    });
  }

  Future<void> setApiBase(String value) async {
    final cleaned = value.trim().replaceAll(RegExp(r'/+$'), '');
    apiBase = cleaned.isEmpty ? kDefaultApiBase : cleaned;
    notifyListeners();
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('apiBase', apiBase);
  }
}

// ── Sesión ─────────────────────────────────────────────────────────────────

class AuthState extends ChangeNotifier {
  final ApiClient api;
  final _storage = const FlutterSecureStorage();

  bool ready = false;
  String? apiKey;
  Map<String, dynamic>? user;
  String notice = ''; // aviso post-logout ("tu sesión expiró…")

  AuthState(this.api);

  Future<void> hydrate() async {
    try {
      final key = await _storage.read(key: _kSecureKey);
      if (key != null && key.isNotEmpty) {
        apiKey = key;
        await refreshMe(); // valida la key; un 401 dispara sessionExpired
      }
    } catch (_) {
      // Keystore no disponible: se queda deslogueado
    }
    ready = true;
    notifyListeners();
  }

  Future<void> refreshMe() async {
    try {
      final data = await api.get('/api/auth/me');
      if (data is Map && data['user'] is Map) {
        user = Map<String, dynamic>.from(data['user'] as Map);
        notifyListeners();
      }
    } on ApiException {
      // el 401 ya deslogueó vía onUnauthorized; otros errores no tumban la sesión
    }
  }

  Future<void> _adoptLogin(Map<String, dynamic> data) async {
    final key = data['api_key'];
    if (key is! String || key.isEmpty) {
      throw ApiException('El servidor no entregó credenciales para la app');
    }
    await _storage.write(key: _kSecureKey, value: key);
    apiKey = key;
    if (data['user'] is Map) {
      user = Map<String, dynamic>.from(data['user'] as Map);
    }
    notice = '';
    notifyListeners();
    unawaited(refreshMe()); // trae plan y settings completos
  }

  Future<void> login(String email, String password) async {
    final data = await api.post('/api/auth/login', auth: false, body: {
      'email': email,
      'password': password,
      'issue_api_key': true,
      'key_name': _kMobileKeyName,
    });
    await _adoptLogin(Map<String, dynamic>.from(data as Map));
  }

  Future<void> register({
    required String firstName,
    required String lastName,
    required String email,
    required String password,
  }) async {
    await api.post('/api/auth/register', auth: false, body: {
      'first_name': firstName,
      'last_name': lastName,
      'email': email,
      'password': password,
    });
    // El registro entrega cookie de sesión (web); la app necesita su API key,
    // así que encadena un login normal con las mismas credenciales.
    await login(email, password);
  }

  Future<void> loginWithProvider(String provider) async {
    final result = await oauthAuthorize(provider, api.base);
    final data = await api.post('/api/auth/oauth/exchange', auth: false, body: {
      'code': result.code,
      'code_verifier': result.verifier,
      'issue_api_key': true,
      'key_name': _kMobileKeyName,
    });
    await _adoptLogin(Map<String, dynamic>.from(data as Map));
  }

  /// La regeneración desde Cuenta devuelve una key nueva: se adopta al instante.
  Future<void> adoptApiKey(String rawKey) async {
    await _storage.write(key: _kSecureKey, value: rawKey);
    apiKey = rawKey;
    notifyListeners();
  }

  void sessionExpired() {
    logout(withNotice: 'Tu sesión expiró. Inicia sesión de nuevo.');
  }

  Future<void> logout({String withNotice = ''}) async {
    try {
      await _storage.delete(key: _kSecureKey);
    } catch (_) {
      // ya borrada
    }
    apiKey = null;
    user = null;
    notice = withNotice;
    notifyListeners();
  }

  void clearNotice() {
    if (notice.isNotEmpty) {
      notice = '';
      notifyListeners();
    }
  }
}

// ── Chat ───────────────────────────────────────────────────────────────────

class ChatMessage {
  final String role;
  String content;
  List<dynamic>? sources;
  ChatMessage(this.role, this.content, {this.sources});
}

class ChatState extends ChangeNotifier {
  final ApiClient api;
  final AuthState auth;

  List<String> models = [];
  String model = '';
  String? conversationId;
  String? title;
  List<ChatMessage> messages = [];
  bool streaming = false;
  bool loadingMessages = false;
  String error = '';
  bool webSearch = false;
  String draft = ''; // texto del compositor (sobrevive al cambio de pestaña)

  ChatStreamHandle? _handle;
  bool _hadHistory = false; // la conversación ya existía (no pedir auto-título)

  ChatState(this.api, this.auth);

  Future<void> loadModels() async {
    try {
      final res = await api.get('/v1/models');
      final list = (res is Map ? res['data'] : null);
      final ids = <String>[];
      if (list is List) {
        for (final m in list) {
          final id = (m is Map ? m['id'] : null);
          if (id is String && !id.startsWith('error:')) ids.add(id);
        }
      }
      models = ids;
      if (model.isEmpty || !ids.contains(model)) {
        model = ids.isNotEmpty ? ids.first : '';
      }
    } catch (_) {
      models = [];
    }
    notifyListeners();
  }

  void setModel(String value) {
    model = value;
    notifyListeners();
  }

  void setWebSearch(bool value) {
    webSearch = value;
    notifyListeners();
  }

  void setDraft(String value) {
    draft = value; // sin notify: el TextField ya refleja el valor
  }

  void clearError() {
    error = '';
    notifyListeners();
  }

  void newChat() {
    stop();
    conversationId = null;
    title = null;
    messages = [];
    error = '';
    _hadHistory = false;
    notifyListeners();
  }

  Future<void> openConversation(String id, {String? withTitle}) async {
    stop();
    conversationId = id;
    title = withTitle;
    messages = [];
    error = '';
    loadingMessages = true;
    _hadHistory = true;
    notifyListeners();
    try {
      final res = await api.get('/api/conversations/$id/messages');
      final msgs = <ChatMessage>[];
      if (res is Map && res['messages'] is List) {
        for (final m in res['messages'] as List) {
          if (m is Map && m['role'] is String) {
            msgs.add(ChatMessage(m['role'] as String, (m['content'] ?? '') as String));
          }
        }
      }
      messages = msgs;
      final conv = (res is Map ? res['conversation'] : null);
      if (conv is Map && conv['title'] is String) title = conv['title'] as String;
    } on ApiException catch (err) {
      error = err.message;
    } finally {
      loadingMessages = false;
      notifyListeners();
    }
  }

  String _randomUuid() {
    // UUID v4 (mismo rol que crypto.randomUUID() en la web)
    final r = List<int>.generate(16, (_) => _secureRandom.nextInt(256));
    r[6] = (r[6] & 0x0F) | 0x40;
    r[8] = (r[8] & 0x3F) | 0x80;
    String hex(int start, int end) => r
        .sublist(start, end)
        .map((b) => b.toRadixString(16).padLeft(2, '0'))
        .join();
    return '${hex(0, 4)}-${hex(4, 6)}-${hex(6, 8)}-${hex(8, 10)}-${hex(10, 16)}';
  }

  Future<void> send(String text) async {
    final trimmed = text.trim();
    if (streaming || trimmed.isEmpty) return;
    if (model.isEmpty) {
      error = 'No hay modelos disponibles ahora mismo. Reintenta en un momento.';
      notifyListeners();
      return;
    }

    final convId = conversationId ?? _randomUuid();
    final isFirstExchange = messages.isEmpty && !_hadHistory;
    final history = [
      ...messages.map((m) => {'role': m.role, 'content': m.content}),
      {'role': 'user', 'content': trimmed},
    ];

    conversationId = convId;
    messages = [...messages, ChatMessage('user', trimmed), ChatMessage('assistant', '')];
    streaming = true;
    error = '';
    draft = '';
    notifyListeners();

    _handle = streamChatCompletion(
      base: api.base,
      token: auth.apiKey,
      model: model,
      messages: history,
      conversationId: convId,
      webSearch: webSearch,
      onDelta: (delta) {
        messages.last.content += delta;
        notifyListeners();
      },
      onSources: (sources) {
        messages.last.sources = sources;
        notifyListeners();
      },
      onDone: () {
        streaming = false;
        _handle = null;
        notifyListeners();
        if (isFirstExchange) _generateTitle(convId);
      },
      onError: (err) {
        // Si no llegó nada, se retira la burbuja vacía del asistente
        if (messages.isNotEmpty &&
            messages.last.role == 'assistant' &&
            messages.last.content.isEmpty) {
          messages = messages.sublist(0, messages.length - 1);
        }
        streaming = false;
        _handle = null;
        error = err is ApiException ? err.message : 'Se perdió la conexión con el servidor';
        notifyListeners();
      },
    );
  }

  void stop() {
    _handle?.cancel();
    _handle = null;
    if (streaming) {
      streaming = false;
      notifyListeners();
    }
  }

  Future<void> _generateTitle(String convId) async {
    try {
      final res = await api.post('/api/conversations/$convId/generate-title');
      if (conversationId == convId && res is Map && res['title'] is String) {
        title = res['title'] as String;
        notifyListeners();
      }
    } catch (_) {
      // sin título automático: no es crítico
    }
  }
}

final Random _secureRandom = Random.secure();

File: apps/mobile/src/state.js
// state.js — estado global con React Context:
//   PrefsContext → tema y servidor (AsyncStorage; nunca credenciales)
//   AuthContext  → sesión (API key "Lixbon Mobile" en el Keystore vía
//                  expo-secure-store) — se rota en cada login
//   ChatContext  → modelos, conversación activa y streaming
import AsyncStorage from '@react-native-async-storage/async-storage';
import * as Crypto from 'expo-crypto';
import * as SecureStore from 'expo-secure-store';
import React, {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';

import { ApiClient, ApiException } from './api';
import { oauthAuthorize } from './oauth';
import { registerPushToken } from './push';
import { streamChatCompletion } from './sse';

export const DEFAULT_API_BASE = 'https://lixbon.com';
const SECURE_KEY = 'lixbon_api_key';
const MOBILE_KEY_NAME = 'Lixbon Mobile';

const ApiContext = createContext(null);
const PrefsContext = createContext(null);
const AuthContext = createContext(null);
const ChatContext = createContext(null);

export const useApi = () => useContext(ApiContext);
export const usePrefs = () => useContext(PrefsContext);
export const useAuth = () => useContext(AuthContext);
export const useChat = () => useContext(ChatContext);

export function AppState({ children }) {
  // Valores vivos que el ApiClient lee en cada petición (evita closures viejas).
  const live = useRef({ apiBase: DEFAULT_API_BASE, apiKey: null, onUnauthorized: () => {} }).current;
  const api = useMemo(
    () =>
      new ApiClient({
        getBase: () => live.apiBase,
        getToken: () => live.apiKey,
        onUnauthorized: () => live.onUnauthorized(),
      }),
    [live],
  );

  // ── Preferencias locales ─────────────────────────────────────────────────
  const [themeMode, setThemeModeState] = useState('system'); // system | light | dark
  const [apiBase, setApiBaseState] = useState(DEFAULT_API_BASE);
  const [prefsReady, setPrefsReady] = useState(false);

  useEffect(() => {
    (async () => {
      try {
        const [mode, base] = await Promise.all([
          AsyncStorage.getItem('themeMode'),
          AsyncStorage.getItem('apiBase'),
        ]);
        if (mode === 'light' || mode === 'dark') setThemeModeState(mode);
        if (base) {
          live.apiBase = base;
          setApiBaseState(base);
        }
      } catch {
        // sin preferencias guardadas
      }
      setPrefsReady(true);
    })();
  }, [live]);

  const setThemeMode = useCallback((mode) => {
    setThemeModeState(mode);
    AsyncStorage.setItem('themeMode', mode).catch(() => {});
  }, []);

  const setApiBase = useCallback(
    (value) => {
      const cleaned = value.trim().replace(/\/+$/, '') || DEFAULT_API_BASE;
      live.apiBase = cleaned;
      setApiBaseState(cleaned);
      AsyncStorage.setItem('apiBase', cleaned).catch(() => {});
    },
    [live],
  );

  const prefs = useMemo(
    () => ({ themeMode, apiBase, setThemeMode, setApiBase }),
    [themeMode, apiBase, setThemeMode, setApiBase],
  );

  // ── Sesión ───────────────────────────────────────────────────────────────
  const [authReady, setAuthReady] = useState(false);
  const [apiKey, setApiKey] = useState(null);
  const [user, setUser] = useState(null);
  const [notice, setNotice] = useState(''); // aviso post-logout ("tu sesión expiró…")

  const adoptKey = useCallback(
    (key) => {
      live.apiKey = key;
      setApiKey(key);
    },
    [live],
  );

  const refreshMe = useCallback(async () => {
    try {
      const data = await api.get('/api/auth/me');
      if (data?.user && typeof data.user === 'object') setUser(data.user);
    } catch {
      // el 401 ya deslogueó vía onUnauthorized; otros errores no tumban la sesión
    }
  }, [api]);

  const logout = useCallback(
    async (withNotice = '') => {
      try {
        await SecureStore.deleteItemAsync(SECURE_KEY);
      } catch {
        // ya borrada
      }
      live.apiKey = null;
      setApiKey(null);
      setUser(null);
      setNotice(withNotice);
    },
    [live],
  );

  live.onUnauthorized = () => logout('Tu sesión expiró. Inicia sesión de nuevo.');

  useEffect(() => {
    (async () => {
      try {
        const key = await SecureStore.getItemAsync(SECURE_KEY);
        if (key) {
          adoptKey(key);
          await refreshMe(); // valida la key; un 401 dispara el logout
        }
      } catch {
        // Keystore no disponible: se queda deslogueado
      }
      setAuthReady(true);
    })();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const adoptLogin = useCallback(
    async (data) => {
      const key = data?.api_key;
      if (typeof key !== 'string' || !key) {
        throw new ApiException('El servidor no entregó credenciales para la app');
      }
      await SecureStore.setItemAsync(SECURE_KEY, key);
      adoptKey(key);
      if (data.user && typeof data.user === 'object') setUser(data.user);
      setNotice('');
      refreshMe(); // trae plan y settings completos
    },
    [adoptKey, refreshMe],
  );

  const login = useCallback(
    async (email, password) => {
      const data = await api.post(
        '/api/auth/login',
        { email, password, issue_api_key: true, key_name: MOBILE_KEY_NAME },
        { auth: false },
      );
      await adoptLogin(data);
    },
    [api, adoptLogin],
  );

  const register = useCallback(
    async ({ firstName, lastName, email, password }) => {
      await api.post(
        '/api/auth/register',
        { first_name: firstName, last_name: lastName, email, password },
        { auth: false },
      );
      // El registro entrega cookie de sesión (web); la app necesita su API key,
      // así que encadena un login normal con las mismas credenciales.
      await login(email, password);
    },
    [api, login],
  );

  const loginWithProvider = useCallback(
    async (provider) => {
      const result = await oauthAuthorize(provider, api.base);
      const data = await api.post(
        '/api/auth/oauth/exchange',
        {
          code: result.code,
          code_verifier: result.verifier,
          issue_api_key: true,
          key_name: MOBILE_KEY_NAME,
        },
        { auth: false },
      );
      await adoptLogin(data);
    },
    [api, adoptLogin],
  );

  /// La regeneración desde Cuenta devuelve una key nueva: se adopta al instante.
  const adoptApiKey = useCallback(
    async (rawKey) => {
      await SecureStore.setItemAsync(SECURE_KEY, rawKey);
      adoptKey(rawKey);
    },
    [adoptKey],
  );

  const clearNotice = useCallback(() => setNotice(''), []);

  // Con sesión activa, registra el push token del dispositivo (avisos de
  // /remote con la app cerrada). Best-effort: sin FCM/permiso es un no-op.
  useEffect(() => {
    if (apiKey) registerPushToken(api);
  }, [apiKey, api]);

  const auth = useMemo(
    () => ({
      ready: authReady,
      apiKey,
      user,
      notice,
      login,
      register,
      loginWithProvider,
      adoptApiKey,
      refreshMe,
      logout,
      clearNotice,
    }),
    [authReady, apiKey, user, notice, login, register, loginWithProvider, adoptApiKey, refreshMe, logout, clearNotice],
  );

  return (
    <ApiContext.Provider value={api}>
      <PrefsContext.Provider value={prefs}>
        <AuthContext.Provider value={auth}>
          <ChatState api={api} apiKey={apiKey} ready={prefsReady && authReady}>
            {children}
          </ChatState>
        </AuthContext.Provider>
      </PrefsContext.Provider>
    </ApiContext.Provider>
  );
}

// ── Chat ─────────────────────────────────────────────────────────────────────

function ChatState({ api, apiKey, ready, children }) {
  const [models, setModels] = useState([]);
  const [model, setModel] = useState('');
  const [conversationId, setConversationId] = useState(null);
  const [title, setTitle] = useState(null);
  const [messages, setMessages] = useState([]);
  const [streaming, setStreaming] = useState(false);
  const [loadingMessages, setLoadingMessages] = useState(false);
  const [error, setError] = useState('');
  const [webSearch, setWebSearch] = useState(false);
  const draftRef = useRef(''); // texto del compositor (sobrevive al cambio de pestaña)

  const handleRef = useRef(null);
  const hadHistoryRef = useRef(false); // la conversación ya existía (no pedir auto-título)
  const messagesRef = useRef([]); // espejo síncrono para send()
  useEffect(() => {
    messagesRef.current = messages;
  }, [messages]);

  const loadModels = useCallback(async () => {
    try {
      const res = await api.get('/v1/models');
      const ids = (Array.isArray(res?.data) ? res.data : [])
        .map((m) => m?.id)
        .filter((id) => typeof id === 'string' && !id.startsWith('error:'));
      setModels(ids);
      setModel((current) => (current && ids.includes(current) ? current : ids[0] || ''));
    } catch {
      setModels([]);
    }
  }, [api]);

  const stop = useCallback(() => {
    handleRef.current?.cancel();
    handleRef.current = null;
    setStreaming(false);
  }, []);

  const newChat = useCallback(() => {
    stop();
    setConversationId(null);
    setTitle(null);
    setMessages([]);
    setError('');
    hadHistoryRef.current = false;
  }, [stop]);

  const openConversation = useCallback(
    async (id, withTitle = null) => {
      stop();
      setConversationId(id);
      setTitle(withTitle);
      setMessages([]);
      setError('');
      setLoadingMessages(true);
      hadHistoryRef.current = true;
      try {
        const res = await api.get(`/api/conversations/${id}/messages`);
        const msgs = (Array.isArray(res?.messages) ? res.messages : [])
          .filter((m) => m && typeof m.role === 'string')
          .map((m) => ({ role: m.role, content: m.content || '' }));
        setMessages(msgs);
        if (typeof res?.conversation?.title === 'string') setTitle(res.conversation.title);
      } catch (err) {
        setError(err instanceof ApiException ? err.message : 'Sin conexión con el servidor');
      } finally {
        setLoadingMessages(false);
      }
    },
    [api, stop],
  );

  const generateTitle = useCallback(
    async (convId) => {
      try {
        const res = await api.post(`/api/conversations/${convId}/generate-title`);
        if (typeof res?.title === 'string') {
          setConversationId((current) => {
            if (current === convId) setTitle(res.title);
            return current;
          });
        }
      } catch {
        // sin título automático: no es crítico
      }
    },
    [api],
  );

  const send = useCallback(
    (text) => {
      const trimmed = text.trim();
      if (handleRef.current || !trimmed) return;
      if (!model) {
        setError('No hay modelos disponibles ahora mismo. Reintenta en un momento.');
        return;
      }

      const current = messagesRef.current;
      const convId = conversationId || Crypto.randomUUID();
      const isFirstExchange = current.length === 0 && !hadHistoryRef.current;
      const history = [
        ...current.map((m) => ({ role: m.role, content: m.content })),
        { role: 'user', content: trimmed },
      ];

      setConversationId(convId);
      setMessages([
        ...current,
        { role: 'user', content: trimmed },
        { role: 'assistant', content: '' },
      ]);
      setStreaming(true);
      setError('');
      draftRef.current = '';

      handleRef.current = streamChatCompletion({
        base: api.base,
        token: apiKey,
        model,
        messages: history,
        conversationId: convId,
        webSearch,
        onDelta: (delta) => {
          setMessages((msgs) => {
            const last = msgs[msgs.length - 1];
            return [...msgs.slice(0, -1), { ...last, content: last.content + delta }];
          });
        },
        onSources: (sources) => {
          setMessages((msgs) => {
            const last = msgs[msgs.length - 1];
            return [...msgs.slice(0, -1), { ...last, sources }];
          });
        },
        onDone: () => {
          handleRef.current = null;
          setStreaming(false);
          if (isFirstExchange) generateTitle(convId);
        },
        onError: (err) => {
          handleRef.current = null;
          // Si no llegó nada, se retira la burbuja vacía del asistente
          setMessages((msgs) => {
            const last = msgs[msgs.length - 1];
            return last && last.role === 'assistant' && !last.content ? msgs.slice(0, -1) : msgs;
          });
          setStreaming(false);
          setError(
            err instanceof ApiException ? err.message : 'Se perdió la conexión con el servidor',
          );
        },
      });
    },
    [api, apiKey, model, conversationId, webSearch, generateTitle],
  );

  // Al cerrar sesión se descarta la conversación en curso.
  useEffect(() => {
    if (ready && !apiKey) newChat();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [apiKey, ready]);

  const chat = useMemo(
    () => ({
      models,
      model,
      conversationId,
      title,
      messages,
      streaming,
      loadingMessages,
      error,
      webSearch,
      draftRef,
      loadModels,
      setModel,
      setWebSearch,
      clearError: () => setError(''),
      newChat,
      openConversation,
      send,
      stop,
    }),
    [models, model, conversationId, title, messages, streaming, loadingMessages, error, webSearch, loadModels, newChat, openConversation, send, stop],
  );

  return <ChatContext.Provider value={chat}>{children}</ChatContext.Provider>;
}
